File: ruby/pythagorean-triplet/pythagorean_triplet.rb
require 'pry'
# comment
class Triplet
  def initialize(*args)
    @args = *args.flatten
  end

  def sum
    @args.sum
  end

  def product
    @args.reduce(&:*)
  end

  def pythagorean?
    @args[0]**2 + @args[1]**2 == @args[2]**2
  end

  class << self
    def where(input)
      @min_factor = input.fetch(:min_factor, 1)
      @max_factor = input.fetch(:max_factor, 1)
      @sum = input.fetch(:sum, nil)

      return pythagorean_triplets unless @sum

      pythagorean_triplets.select{ |t| t.sum == @sum }
    end

    private

    def pythagorean_triplets
      (@min_factor..@max_factor)
        .to_a
        .combination(3)
        .select { |arr| Triplet.new(arr).pythagorean? }
        .map { |arr| Triplet.new(arr) }
    end
  end
end

File: javascript/pythagorean-triplet/pythagorean-triplet.js
//
// This is only a SKELETON file for the 'Pythagorean Triplet' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export function triplets({ minFactor, maxFactor, sum }) {
  // let minFactor = 1
  // let maxFactor = sum - 2

  let range = []
  for (let i = minFactor; i <= maxFactor; i++) {
    range.push(i)
  }
  console.log(range)

  // generate rtipleets from range
  let combinations = []
  for (let i = 0; i < range.length -2; i++) {
    for (let j = i + 1; j < range.length -1; j++) {
      for (let k = j + 1; k < range.length; k++) {
        combinations.push( [range[i], range[j], range[k]])
      }
    }
  }
  console.log(combinations)
}

class Triplet {
  constructor(a, b, c) {
    this.numbers = [a, b, c]
  }

  sum = () => {
    return this
            .numbers
            .reduce( (acc, element) => {return(acc + element)}, 0)
  }

  pythagorean = () => {
    return (
      this.numbers[0]**2 + this.numbers[1]**2 == this.numbers[2]**2
      )
  }

  get toArray() {
    return this.numbers
  }
}
